core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java

```diff
@@ -473,6 +479,107 @@ public TrimResult trimFields(
     return result(newCalc, mapping, calc);
   }
 
+  /**
+   * Shuttle that finds all {@link TableScan}`s inside a given {@link RelNode}.
+   */
+  private static class TableScanCollector extends RelHomogeneousShuttle {
+    private ImmutableSet.Builder<List<String>> builder = ImmutableSet.builder();
+
+    /** Qualified names. */
+    Set<List<String>> tables() {
+      return builder.build();
+    }
+
+    @Override public RelNode visit(TableScan scan) {
+      builder.add(scan.getTable().getQualifiedName());
+      return super.visit(scan);
+    }
+  }
+
+  /**
+   * Shuttle that finds all {@link TableScan}`s inside a given {@link RexNode}.
+   */
+  private static class InputTablesVisitor extends RexVisitorImpl<Void> {
```

Comment:
Should there a boolean flag to `RexVisitorImpl`'s constructor that says whether to visit subqueries?

Reply:
if you talk about `RexVisitorImpl(boolean deep)` i\`l fix it